### PR TITLE
[FE] 4.1.7 채팅 타임라인 및 메시지별 워크플로우 매핑

### PR DESCRIPTION
**Pull request by kang-heewon**
4.1.7 spec 기반 구현 완료

---

<a href="https://app.devin.ai/review/ajou-2026-1-capstone-5/ostone/pull/219" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

## 릴리즈 노트

* **새로운 기능**
  * 타임스탬프가 보이는 클릭 가능한 채팅 타임라인 추가
  * 상태 머신을 시각화하는 새로운 그래프 및 사이드 패널 추가
  * 메시지 선택에 따른 하이라이트 동기화 및 사이드 패널 연동
  * 결정 로그 UI 개선(상태 전환, 결정 배지, 신뢰도 표시)

* **개선사항**
  * 실행 상세 정보에서 정책/위험 히트와 슬롯 정보 가시성 향상
  * 페이지 상태 전달을 단일 상태 객체로 통합하여 구조 단순화

* **테스트**
  * 컴포넌트별 테스트 대폭 보강 및 시나리오 커버리지 확장

[Review Change Stack](https://app.coderabbit.ai/change-stack/ajou-2026-1-capstone-5/ostone/pull/219)

### CODE REVIEW

File: frontend/src/features/chat-workflow/ui/index.ts
[the remark this answers is not in the document]
Comment (kang-heewon) [reply frontend/src/features/chat-workflow/ui/index.ts:1]: Thanks for the review. This is an informational finding — the deprecation pattern (keeping the file with an inline type) is intentional per the plan (Task 11). The old test continues to pass against the retained .tsx file, confirming no consumers break.

File: frontend/src/pages/chat-demo/ui/ChatWorkflowDemoPage.tsx [hunk 1]
```diff
@@ -49,40 +46,51 @@ export function ChatWorkflowDemoPage({
...
               background: 'var(--paper-2)',
               borderRadius: 'var(--r-2)',
               overflow: 'hidden',
             }}
           >
-            <ExecutionDetailPanel status={workflow.status} context={workflow.context} />
+            {workflow ? (
+              <SidePanel
+                workflow={workflow}
+                execution={execution!}
```
[the remark this answers is not in the document]
Comment (kang-heewon) [reply frontend/src/pages/chat-demo/ui/ChatWorkflowDemoPage.tsx]: Thanks for confirming. This is an informational finding and matches the intended behavior — the non-null assertion is safe because execution is a required field when response exists. No change needed.

File: frontend/src/pages/chat-demo/ui/ChatWorkflowDemoPage.tsx [hunk 2]
```diff
@@ -1,35 +1,32 @@
...
 
-export function ChatWorkflowDemoPage({
-  domainPack,
-  scenario,
-  messages,
-  workflow,
-  decisionLog,
-}: ChatWorkflowDemoPageProps) {
+export function ChatWorkflowDemoPage({ state }: ChatWorkflowDemoPageProps) {
+  const [selectedMessageId, setSelectedMessageId] = useState<string | null>(null);
```
[the remark this answers is not in the document]
Comment (kang-heewon) [reply frontend/src/pages/chat-demo/ui/ChatWorkflowDemoPage.tsx]: Thanks for the review. This is intentional — the DemoChatWorkflowState.selectedMessageId is part of the response model defined in 4.1.5 API, but the page uses local useState for interactive state management (component-controlled pattern). The field exists in the type for potential future use when loading from real API data.

File: frontend/src/features/chat-workflow/ui/StateMachineGraph.tsx [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+  return (
+    <div data-testid="graph-container" className="w-full overflow-hidden">
+      <svg width={svgWidth(states.length)} height={SVG_H} className="block">
+        <defs>
+          <marker id="arr" markerWidth="10" markerHeight="7" refX="10" refY="3.5" orient="auto">
+            <polygon points="0 0, 10 3.5, 0 7" fill="#9ca3af" />
+          </marker>
+          <marker id="arr-blue" markerWidth="10" markerHeight="7" refX="10" refY="3.5" orient="auto">
+            <polygon points="0 0, 10 3.5, 0 7" fill="#3b82f6" />
+          </marker>
```
[the remark this answers is not in the document]
Comment (kang-heewon) [reply frontend/src/features/chat-workflow/ui/StateMachineGraph.tsx:58]: Good catch. The marker IDs are global to the SVG but the component is rendered as a single instance in the side panel — multiple instances on the same page is outside current scope. If reuse is needed later, we can use React.useId() or unique prefix prop. Noted for future refactoring.

File: frontend/src/features/chat-workflow/model/chatWorkflow.types.ts [hunk 4]
```diff
@@ -1,36 +1,111 @@
...
+  status: string;
+  startedAt: string;
+  completedAt?: string;
 }
 
-export interface ChatMessage {
+export interface DemoChatMessage {
   id: string;
-  role: string;
+  role: 'user' | 'assistant';
```
[the remark this answers is not in the document]
Comment (kang-heewon) [reply frontend/src/features/chat-workflow/model/chatWorkflow.types.ts:54]: Acknowledged. The role union 'user' | 'assistant' matches the 4.1.5 mock API fixture which only produces these two roles. If a future real backend returns 'system' messages, the type can be widened. This is a deliberate narrowing for the demo scope.

File: frontend/src/pages/chat-demo/ui/ChatWorkflowDemoPage.tsx [hunk 5]
```diff
@@ -1,35 +1,32 @@
...
+
+  const response = state.response;
+  const messages: DemoChatMessage[] = response?.messages ?? [];
+  const execution = response?.execution ?? null;
+  const decisionLogs: DemoDecisionLogEntry[] = response?.decisionLogs ?? [];
+  const domainPack = response?.domainPack ?? null;
+  const workflow = response?.workflow ?? null;
```
[the remark this answers is not in the document]
Comment (kang-heewon) [reply frontend/src/pages/chat-demo/ui/ChatWorkflowDemoPage.tsx:24]: This was already fixed in commit 68246c5 — ChatWorkflowDemoPage now has loading/error early returns. Resolving.

File: frontend/src/pages/chat-demo/ui/ChatWorkflowDemoPage.tsx [hunk 6]
```diff
@@ -38,7 +35,7 @@ export function ChatWorkflowDemoPage({
           gap: 'var(--s-4)',
         }}
       >
-        <ChatWorkflowHeader domainPack={domainPack} scenario={scenario} />
+        <ChatWorkflowHeader domainPack={domainPack} />
```
[the remark this answers is not in the document]
Comment (kang-heewon) [reply frontend/src/pages/chat-demo/ui/ChatWorkflowDemoPage.tsx]: Fixed — removed the duplicate ChatWorkflowHeader from ChatWorkflowDemoPage.tsx. The header now only renders inside SidePanel.

File: frontend/src/features/chat-workflow/ui/StateMachineGraph.tsx [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+      highlightedStates.add(log.stateFrom);
+      highlightedStates.add(log.stateTo);
+      matchedTransitionKeys.add(`${log.stateFrom}\u2192${log.stateTo}`);
+    }
+  }
+
+  const hasFilter = selectedMessageId !== null;
+
+  return (
+    <div data-testid="graph-container" className="w-full overflow-hidden">
```
[the remark this answers is not in the document]
Comment (kang-heewon) [reply frontend/src/features/chat-workflow/ui/StateMachineGraph.tsx]: Fixed — changed overflow-hidden to overflow-x-auto on graph-container so the full state machine graph remains scrollable.

File: frontend/src/features/chat-workflow/ui/DecisionLogDrawer.tsx [hunk 8]
```diff
@@ -1,37 +1,128 @@
...
+                    <p style={{ margin: 0, fontSize: '13px', color: '#6b7280' }}>
+                      {entry.reason}
+                    </p>
+                  </li>
+                );
+              })}
             </ul>
           )}
         </div>
       )}
```
[the remark this answers is not in the document]
Comment (kang-heewon) [reply frontend/src/features/chat-workflow/ui/DecisionLogDrawer.tsx:125]: DecisionLogDrawer는 props-driven 컴포넌트로, loading/error 상태는 entries prop의 nullability와 부모(ChatWorkflowDemoPage)에서 처리합니다. entries type이 DemoDecisionLogEntry[]이고 null이 될 수 없으므로 entries 자체가 있으면 표시, 없으면 empty 처리됩니다. 실제 loading/error는 페이지 레벨에서 처리하므로 컴포넌트 레벨에서 별도 분기가 필요하지 않습니다.

File: frontend/src/features/chat-workflow/ui/ExecutionDetailPanel.tsx [hunk 9]
```diff
@@ -1,36 +1,127 @@
...
+          <div data-testid="execution-risks">
+            <span className="mb-1 block text-xs text-muted-foreground">Risk Hits:</span>
+            <div className="space-y-1.5">
+              {execution.riskHits.map((hit) => (
+                <RiskHitRow key={hit.riskId} hit={hit} />
+              ))}
+            </div>
+          </div>
+        )}
+      </div>
```
[the remark this answers is not in the document]
Comment (kang-heewon) [reply frontend/src/features/chat-workflow/ui/ExecutionDetailPanel.tsx:124]: ExecutionDetailPanel도 props-driven 컴포넌트입니다. execution: DemoExecution | null 타입으로 null이면 No execution data를 표시합니다. 에러 상태는 부모 컴포넌트(ChatWorkflowDemoPage)에서 관리하며 API 실패 시 response 자체가 null이 되어 페이지 레벨에서 처리됩니다. 별도 error prop이 없어도 데모 시나리오에서 충분히 안전합니다.

File: frontend/src/features/chat-workflow/ui/StateMachineGraph.test.tsx [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+        transitions={transitions}
+        decisionLogs={decisionLogs}
+        selectedMessageId={null}
+        currentState="INITIAL"
+      />
+    );
+
+    expect(screen.getByTestId('graph-container')).toBeInTheDocument();
+  });
+});
```
[the remark this answers is not in the document]
Comment (kang-heewon) [reply frontend/src/features/chat-workflow/ui/StateMachineGraph.test.tsx:113]: Thanks for the suggestion. These edge cases (empty states, invalid transitions, empty decisionLogs) would add value for production code. Since this is a demo page component and the plan prioritizes core functionality first, we will note these for future enhancement.

File: frontend/src/features/chat-workflow/ui/StateMachineGraph.tsx [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+}: StateMachineGraphProps) {
+  const highlightedStates = new Set<string>();
+  const matchedTransitionKeys = new Set<string>();
+
+  if (selectedMessageId) {
+    const matchedLogs = decisionLogs.filter((log) => log.messageId === selectedMessageId);
+    for (const log of matchedLogs) {
+      highlightedStates.add(log.stateFrom);
+      highlightedStates.add(log.stateTo);
+      matchedTransitionKeys.add(`${log.stateFrom}\u2192${log.stateTo}`);
```
[the remark this answers is not in the document]
Comment (kang-heewon) [reply frontend/src/features/chat-workflow/ui/StateMachineGraph.tsx:43]: Good observation. The unicode is only used as a React key separator and is never parsed back. State names containing would be extremely unlikely, and even if they did, the visual impact would be limited to incorrect highlighting on the graph. Acceptable risk for a demo component.

File: frontend/src/features/chat-workflow/ui/StateMachineGraph.tsx [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+          if (isCurrent) {
+            strokeColor = "#22c55e";
+            textColor = "#22c55e";
+            strokeW = 3;
+            fillColor = "#f0fdf4";
+          } else if (inPath) {
+            strokeColor = "#3b82f6";
+            textColor = "#3b82f6";
+            fillColor = "#eff6ff";
+          }
```
[the remark this answers is not in the document]
Comment (kang-heewon) [reply frontend/src/features/chat-workflow/ui/StateMachineGraph.tsx:122]: Noted. For a demo component that is not part of the design system, hardcoded colors are acceptable per the KISS principle. If this component is promoted to production, we would extract color tokens.

File: frontend/src/features/chat-workflow/ui/StateMachineGraph.tsx [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+                dominantBaseline="central"
+                fontSize="13"
+                fontWeight={isCurrent ? "700" : "500"}
+                fill={textColor}
+                className="select-none"
+              >
+                {state}
+              </text>
+            </g>
+          );
```
[the remark this answers is not in the document]
Comment (kang-heewon) [reply frontend/src/features/chat-workflow/ui/StateMachineGraph.tsx:154]: Good suggestion for accessibility. The component already has data-testid attributes for automated testing. Full ARIA support would be appropriate in a production iteration.

File: frontend/src/pages/chat-demo/ui/ChatWorkflowDemoPage.tsx [hunk 14]
```diff
@@ -1,35 +1,52 @@
...
 
-export function ChatWorkflowDemoPage({
-  domainPack,
-  scenario,
-  messages,
-  workflow,
-  decisionLog,
-}: ChatWorkflowDemoPageProps) {
+export function ChatWorkflowDemoPage({ state }: ChatWorkflowDemoPageProps) {
+  const [selectedMessageId, setSelectedMessageId] = useState<string | null>(null);
```
[the remark this answers is not in the document]
Comment (kang-heewon) [reply frontend/src/pages/chat-demo/ui/ChatWorkflowDemoPage.tsx]: Fixed — `useState<string | null>(state.selectedMessageId)` now uses the state prop as initial value, respecting the state model contract.

File: frontend/src/features/chat-workflow/model/chatWorkflowDemo.mock.ts [hunk 15]
```diff
@@ -1,74 +1,232 @@
...
+    id: 'log-5',
+    step: 5,
+    messageId: 'msg-4',
+    eventType: 'ANSWER_GENERATED',
+    stateFrom: 'DECIDING',
+    stateTo: 'COMPLETED',
+    decision: 'ALLOW',
+    confidence: 92,
+    reason: '환불 완료 안내 생성',
+  },
```
[the remark this answers is not in the document]
Comment (kang-heewon) [reply frontend/src/features/chat-workflow/model/chatWorkflowDemo.mock.ts:208]: Noted. This is an intentional simplification in demo mock data — the RISK_CHECKING→DECIDING transition is treated as an internal/automatic transition that doesn't produce a decision log entry. The mock data represents the visible decision points only.

File: frontend/src/pages/chat-demo/ui/ChatWorkflowDemoPage.tsx [hunk 16]
```diff
@@ -1,35 +1,52 @@
...
+
+  if (error) {
+    return (
+      <OstoneShell active="chat-demo" crumbs={['Chat Workflow Demo']}>
+        <div data-testid="error-state" style={{ padding: 'var(--s-8)', textAlign: 'center', color: 'var(--danger)' }}>
+          Error: {error}
+        </div>
+      </OstoneShell>
+    );
+  }
```
[the remark this answers is not in the document]
Comment (kang-heewon) [reply frontend/src/pages/chat-demo/ui/ChatWorkflowDemoPage.tsx:44]: Already addressed in commit 68246c5 — loading and error states are now handled with early returns and dedicated test coverage.

File: frontend/src/pages/chat-demo/ui/ChatWorkflowDemoPage.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (kang-heewon) [reply frontend/src/pages/chat-demo/ui/ChatWorkflowDemoPage.tsx]: Two points: (1) The selectedMessageId initialization issue was already fixed in this commit — `useState<string | null>(state.selectedMessageId)`. (2) Regarding toast.error() — the existing inline error rendering is intentional for a demo page. Adding `sonner` toast dependency to a demo component violates the KISS principle. The error state has a clear visual display with `data-testid='error-state'` for automated testing. If this component is promoted to production, toast integration would be appropriate.
